File: app/Services/DiarioService.php
<?php

namespace App\Services;

use App\Models\Divida;
use App\Models\PerfilFinanceiro;
use Carbon\CarbonImmutable;

/**
 * O número da tela inicial: "pra quitar até janeiro, seu diário é R$ 23".
 * Spec §3.1.
 *
 * Determinístico, sem LLM:
 *
 *   renda_mensal − contas_fixas − parcelas_do_mês = sobra_mensal
 *   diário = sobra_mensal / dias_do_mês
 *
 * A regra que não pode ser quebrada: quando a sobra é negativa o app não
 * esconde e não mostra "R$ 0". Ele diz quanto falta. Um diário de R$ 0
 * fingiria que dá pra viver sem gastar nada — e a pessoa endividada já
 * sabe que não dá.
 */
class DiarioService
{
    /**
     * @return array{
     *   diario: float,
     *   sobra_mensal: float,
     *   renda_mensal: float,
     *   contas_fixas: float,
     *   parcelas_do_mes: float,
     *   dias_no_mes: int,
     *   fecha: bool,
     *   falta_por_mes: float,
     *   quitacao_em: ?string,
     *   quitacao_label: ?string,
     *   tem_dividas: bool,
     * }
     */
    public function calcular(int $userId, ?CarbonImmutable $referencia = null): array
    {
        $hoje = $referencia?->startOfDay() ?? CarbonImmutable::today();

        $perfil = PerfilFinanceiro::query()->where('user_id', $userId)->first();

        $renda = (float) ($perfil?->renda_mensal ?? 0);
        $fixas = (float) ($perfil?->contas_fixas_estimadas ?? 0);

        $dividas = Divida::query()
            ->where('user_id', $userId)
            ->emAberto()
            ->get();

        $parcelas = $this->parcelasDoMes($dividas);

        $sobra = round($renda - $fixas - $parcelas, 2);
        $diasNoMes = $hoje->daysInMonth;

        $fecha = $sobra > 0;

        $quitacao = $this->previsaoQuitacao($dividas, $hoje);

        return [
            'diario' => $fecha ? round($sobra / $diasNoMes, 2) : 0.0,
            'sobra_mensal' => $sobra,
            'renda_mensal' => round($renda, 2),
            'contas_fixas' => round($fixas, 2),
            'parcelas_do_mes' => round($parcelas, 2),
            'dias_no_mes' => $diasNoMes,
            'fecha' => $fecha,
            // Quanto falta por mês para a conta fechar. Zero quando fecha.
            'falta_por_mes' => $fecha ? 0.0 : round(abs($sobra), 2),
            'quitacao_em' => $quitacao?->toDateString(),
            'quitacao_label' => $quitacao ? $this->rotularMes($quitacao) : null,
            'tem_dividas' => $dividas->isNotEmpty(),
        ];
    }

    /**
     * Soma das parcelas que caem no mês. Uma dívida só entra se ainda tem
     * parcela restante — dívida quitada não pesa no diário.
     *
     * @param \Illuminate\Support\Collection<int, Divida> $dividas
     */
    private function parcelasDoMes($dividas): float
    {
        return (float) $dividas
            ->filter(fn (Divida $d) => $d->parcelas_restantes > 0)
            ->sum(fn (Divida $d) => (float) $d->valor_parcela);
    }

    /**
     * A data de quitação é a mais tardia entre as dívidas (spec §3.1):
     * o app só promete "livre" quando a última parcela cair.
     *
     * @param \Illuminate\Support\Collection<int, Divida> $dividas
     */
    private function previsaoQuitacao($dividas, CarbonImmutable $hoje): ?CarbonImmutable
    {
        $datas = $dividas
            ->map(fn (Divida $d) => $d->previsaoQuitacao($hoje))
            ->filter()
            ->values();

        if ($datas->isEmpty()) {
            return null;
        }

        return $datas->sortDesc()->first();
    }

    /** "janeiro", ou "janeiro de 2028" quando não é o ano corrente nem o próximo. */
    private function rotularMes(CarbonImmutable $data): string
    {
        $meses = [
            1 => 'janeiro', 'fevereiro', 'março', 'abril', 'maio', 'junho',
            'julho', 'agosto', 'setembro', 'outubro', 'novembro', 'dezembro',
        ];

        $nome = $meses[(int) $data->month];

        return $data->year > CarbonImmutable::today()->year + 1
            ? "{$nome} de {$data->year}"
            : $nome;
    }
}

File: lib/services/diario_service.dart
import '../models/divida.dart';
import '../models/perfil_financeiro.dart';

/// O resultado do cálculo do diário.
class ResultadoDiario {
  const ResultadoDiario({
    required this.diario,
    required this.sobraMensal,
    required this.parcelasDoMes,
    required this.contasFixas,
    required this.diasNoMes,
    required this.fecha,
    required this.faltaPorMes,
    required this.temDividas,
    this.quitacaoEm,
    this.quitacaoLabel,
  });

  final double diario;
  final double sobraMensal;
  final double parcelasDoMes;
  final double contasFixas;
  final int diasNoMes;

  /// false quando a sobra é zero ou negativa.
  final bool fecha;

  /// Quanto falta por mês para a conta fechar. Zero quando fecha.
  final double faltaPorMes;

  final bool temDividas;
  final DateTime? quitacaoEm;
  final String? quitacaoLabel;
}

/// O número da tela inicial: "pra quitar até janeiro, seu diário é R$ 23".
///
/// Determinístico, sem IA:
///   renda − fixas − parcelas = sobra
///   sobra ÷ dias do mês = diário
///
/// A regra que não pode ser quebrada: **sobra negativa não vira diário
/// R$ 0**. Mostrar zero fingiria que dá pra viver sem gastar nada, e quem
/// está endividado já sabe que não dá. O app diz quanto falta.
class DiarioService {
  const DiarioService();

  ResultadoDiario calcular({
    required PerfilFinanceiro? perfil,
    required List<Divida> dividas,
    DateTime? referencia,
  }) {
    final hoje = referencia ?? DateTime.now();
    final emAberto = dividas.where((d) => !d.estaQuitada).toList();

    final renda = perfil?.rendaMensal ?? 0;
    final fixas = perfil?.contasFixasEstimadas ?? 0;

    // Dívida sem parcela restante não pesa no diário.
    final parcelas = emAberto
        .where((d) => d.parcelasRestantes > 0)
        .fold<double>(0, (soma, d) => soma + d.valorParcela);

    final sobra = _duasCasas(renda - fixas - parcelas);
    final diasNoMes = DateTime(hoje.year, hoje.month + 1, 0).day;

    // Zero é o limite: não sobrou nada para o dia a dia, então não fecha.
    final fecha = sobra > 0;

    final quitacao = _previsaoQuitacao(emAberto, hoje);

    return ResultadoDiario(
      diario: fecha ? _duasCasas(sobra / diasNoMes) : 0,
      sobraMensal: sobra,
      parcelasDoMes: _duasCasas(parcelas),
      contasFixas: _duasCasas(fixas),
      diasNoMes: diasNoMes,
      fecha: fecha,
      faltaPorMes: fecha ? 0 : _duasCasas(sobra.abs()),
      temDividas: emAberto.isNotEmpty,
      quitacaoEm: quitacao,
      quitacaoLabel: quitacao == null ? null : _rotularMes(quitacao),
    );
  }

  /// A data de quitação é a **mais tardia** entre as dívidas: o app só
  /// promete "livre" quando a última parcela cair.
  DateTime? _previsaoQuitacao(List<Divida> dividas, DateTime hoje) {
    final datas = dividas
        .map((d) => d.previsaoQuitacao(hoje))
        .whereType<DateTime>()
        .toList();

    if (datas.isEmpty) return null;

    datas.sort();
    return datas.last;
  }

  static const _meses = [
    'janeiro', 'fevereiro', 'março', 'abril', 'maio', 'junho',
    'julho', 'agosto', 'setembro', 'outubro', 'novembro', 'dezembro',
  ];

  String _rotularMes(DateTime data) {
    final nome = _meses[data.month - 1];
    return data.year > DateTime.now().year + 1
        ? '$nome de ${data.year}'
        : nome;
  }

  double _duasCasas(double v) => (v * 100).round() / 100;
}
